extra/sodium/secure-memory/secure-memory.factor

```diff
@@ -0,0 +1,44 @@
+! Copyright (C) 2020 Alexander Ilin.
+! See http://factorcode.org/license.txt for BSD license.
+USING:
+    accessors alien continuations destructors kernel libc
+    sodium sodium.ffi
+;
+
+IN: sodium.secure-memory
+
+TUPLE: secure-memory < disposable
+    size underlying ;
+
+: new-secure-memory ( size -- obj )
+    dup sodium-malloc secure-memory new-disposable
+    swap >>underlying swap >>size ;
+
+: allow-no-access ( secure-memory -- )
+    check-disposed underlying>> sodium_mprotect_noaccess check0 ;
+
+: allow-read-access ( secure-memory -- )
+    check-disposed underlying>> sodium_mprotect_readonly check0 ;
+
+: with-read-access ( ..a secure-memory quot: ( ..a secure-memory -- ..b ) -- ..b )
+    over dup allow-read-access [ allow-no-access ] curry finally ; inline
+
+: allow-write-access ( secure-memory -- )
+    check-disposed underlying>> sodium_mprotect_readwrite check0 ;
+
+: with-write-access ( ..a secure-memory quot: ( ..a secure-memory -- ..b ) -- ..b )
+    over dup allow-write-access [ allow-no-access ] curry finally ; inline
+
+: with-new-secure-memory ( ..a size quot: ( ..a secure-memory -- ..b ) -- ..b secure-memory )
+    [ new-secure-memory ] dip with-write-access ; inline
+
+M: secure-memory dispose* ( disposable -- )
+    underlying>> sodium_free ;
```

Comment:
Should you change ``underlying>>`` to be ``f``?  like this:

```
[ sodium_free f ] change-underlying
```

Reply:
Not sure if that's necessary, since the `disposed` will be set to `t`, and the object won't be used anymore. But I'll do it.